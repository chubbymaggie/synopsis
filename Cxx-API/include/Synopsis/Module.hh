//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#ifndef _Synopsis_Module_hh
#define _Synopsis_Module_hh

#include <Synopsis/Object.hh>

namespace Synopsis
{

class Module : public Object
{
public:
  Module(const Object &o) : Object(o) {}
  std::string name() const { return PyModule_GetName(my_impl);}
  std::string filename() const { return PyModule_GetFilename(my_impl);}
  Dict dict() const;

  static Module import(const std::string &name) { return Object::import(name);}
  static Module define(const std::string &name, PyMethodDef *methods);
private:
  Module(PyObject *m) : Object(m) {}
};

inline Module Module::define(const std::string &name, PyMethodDef *methods)
{
  PyObject *m = Py_InitModule(const_cast<char *>(name.c_str()), methods);
  Py_INCREF(m);
  return Module(m);
}

inline Dict Module::dict() const
{
  PyObject *d = PyModule_GetDict(my_impl);
  Py_INCREF(d);
  return Object(d);
}

}

#endif
