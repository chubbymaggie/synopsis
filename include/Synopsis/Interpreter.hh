// $Id: Interpreter.hh,v 1.1 2004/01/09 20:03:25 stefan Exp $
//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#ifndef _Synopsis_Interpreter_hh
#define _Synopsis_Interpreter_hh

#include <Synopsis/Object.hh>
#include <string>

namespace Synopsis
{

class Interpreter
{
public:

   enum Mode { EVAL = Py_eval_input,
               FILE = Py_file_input,
               SINGLE = Py_single_input};
   
   Interpreter() {}
   ~Interpreter() {}

   Object run_string(const std::string &, Mode, Object, Object);
   Object run_file(const std::string &, Mode, Object, Object);
private:
};

inline Object Interpreter::run_string(const std::string &code,
                                      Mode m, Object globals, Object locals)
{
   Object retn = PyRun_String(const_cast<char *>(code.c_str()), m,
                              globals.my_impl, locals.my_impl);
   if (!retn) PyErr_Print();
   return retn;
}

inline Object Interpreter::run_file(const std::string &script,
                                    Mode m, Object globals, Object locals)
{
   ::FILE *file = fopen(script.c_str(), "r");
   Object retn = PyRun_File(file, const_cast<char *>(script.c_str()),
                            m, globals.my_impl, locals.my_impl);
   fclose(file);
   if (!retn) PyErr_Print();
   return retn;
}

}

#endif
