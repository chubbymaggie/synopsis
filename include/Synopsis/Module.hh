// $Id: Module.hh,v 1.3 2004/01/24 04:44:07 stefan Exp $
//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#ifndef _Synopsis_Module_hh
#define _Synopsis_Module_hh

#include <Synopsis/Object.hh>

namespace Synopsis
{

class Module : public Object
{
public:
  Module(const std::string &);
  std::string name() const { return PyModule_GetName(my_impl);}
  std::string filename() const { return PyModule_GetFilename(my_impl);}
  Dict dict() const;
};

inline Module::Module(const std::string &name)
  : Object(PyImport_ImportModule(const_cast<char *>(name.c_str())))
{
  Py_INCREF(my_impl);
}

inline Dict Module::dict() const
{
  PyObject *d = PyModule_GetDict(my_impl);
  Py_INCREF(d);
  return Object(d);
}

}

#endif
