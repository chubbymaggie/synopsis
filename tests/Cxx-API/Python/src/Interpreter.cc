#include <iostream>
#include <Synopsis/Interpreter.hh>
#include <Synopsis/Module.hh>
#include <string>
#include <cstdio>
#include <iostream>

using namespace Synopsis;

void print(Dict &d)
{
  List keys = d.keys();
  for (int i = 0; i != keys.size(); ++i)
  {
     Object key = keys.get(i);
     std::cout << Object::narrow<const char *>(key) << ' '
               << Object::narrow<const char *>(d.get(key).str()) << std::endl;
  }
}

void test1()
{
  std::cout << "=== test1 ===" << std::endl;
  Interpreter interp;
  Module module("__main__");
  Dict global = module.dict();
  Dict local;
  Object retn = interp.run_string("print 'hello world'", Interpreter::FILE,
                                  global, local);
}

void test2()
{
  std::cout << "=== test2 ===" << std::endl;
  Interpreter interp;
  Module module("__main__");
  Dict global = module.dict();
  Dict local;
  Object retn = interp.run_file("script.py", Interpreter::FILE,
                                global, local);
  print(global);
  print(local);  
}

void test3()
{
  std::cout << "=== test3 ===" << std::endl;
  Interpreter interp;
  Module module("__main__");
  Dict global = module.dict();
  Dict local;
  Object retn = interp.run_file("Command.py", Interpreter::FILE,
                                global, local);
  Object type = local.get("Command");
  Tuple args("first", "second", "third");
  Dict kwds;
  kwds.set("input", "foo.h");
  kwds.set("output", "foo.i");
  retn = type();
  retn = type(args);
  retn = type(args, kwds);
  Object method = retn.attr("execute");
  method();
}

int main(int, char **)
{  
  try
  {
    test1();
    test2();
    test3();
  }
  catch (const Interpreter::Exception &)
  {
    PyErr_Print();
  }
  catch (const std::exception &e)
  {
    std::cout << "Error : " << e.what() << std::endl;
  }
}
