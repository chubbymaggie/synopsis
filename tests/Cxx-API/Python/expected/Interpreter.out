=== test1 ===
hello world
=== test2 ===
hello world
=== test3 ===
()
{}
('first', 'second', 'third')
{}
('first', 'second', 'third')
{'input': 'foo.h', 'output': 'foo.i'}
hello world
